erizo/src/erizo/NicerConnection.cpp

```diff
@@ -26,74 +26,643 @@ extern "C" {
+    return 0;
+  }
+  conn->updateIceState(IceState::READY);
+  conn->nicer_->IceContextFinalize(conn->ctx_, pctx);
+
+  return 0;
+}
+
+int NicerConnection::ice_disconnected(void *obj, nr_ice_peer_ctx *pctx) {
+  return 0;
+}
+
+int NicerConnection::msg_recvd(void *obj, nr_ice_peer_ctx *pctx,
+                               nr_ice_media_stream *stream, int component_id,
+                               unsigned char *msg, int len) {
+  NicerConnection *conn = reinterpret_cast<NicerConnection*>(obj);
+  conn->onData(component_id, reinterpret_cast<char*> (msg), static_cast<unsigned int> (len));
+  return 0;
+}
+
+void NicerConnection::trickle_callback(void *arg, nr_ice_ctx *ctx, nr_ice_media_stream *stream,
+                                       int component_id, nr_ice_candidate *candidate) {
+  NicerConnection *conn = reinterpret_cast<NicerConnection*>(arg);
+  conn->onCandidate(stream, component_id, candidate);
+}
+
+NicerConnection::NicerConnection(std::shared_ptr<IOWorker> io_worker, std::shared_ptr<NicerInterface> interface,
+                                 IceConnectionListener *listener, const IceConfig& ice_config)
+    : IceConnection(listener, ice_config),
+      io_worker_{io_worker},
+      nicer_{interface},
+      ice_config_{ice_config},
+      closed_{false},
+      name_{ice_config_.transport_name},
+      ctx_{nullptr},
+      peer_{nullptr},
+      stream_{nullptr},
+      offerer_{!ice_config_.username.empty() && !ice_config_.password.empty()} {
 }
 
 NicerConnection::~NicerConnection() {
   close();
 }
 
+void NicerConnection::async(function<void()> f) {
+  std::weak_ptr<NicerConnection> weak_this = shared_from_this();
+  io_worker_->task([weak_this, f] {
+    if (!weak_this.lock()) {
+      return;
+    }
+    f();
+  });
+}
+
 void NicerConnection::start() {
-  std::string name = "agent";
-  // Create the ICE context
+  async([this] {
+    startSync();
+  });
+  std::future_status status = start_promise_.get_future().wait_for(std::chrono::seconds(5));
+  if (status == std::future_status::timeout) {
+    ELOG_WARN("%s Start timed out", toLog());
+  }
+}
+
+void NicerConnection::startSync() {
   UINT4 flags = NR_ICE_CTX_FLAGS_AGGRESSIVE_NOMINATION;
 
-  nr_ice_ctx_create(const_cast<char *>(name.c_str()), flags, &ctx_);
+  ufrag_ = getNewUfrag();
+  upass_ = getNewPwd();
+  if (ufrag_.empty() || upass_.empty()) {
+    start_promise_.set_value();
+    return;
+  }
+
+  int r = nicer_->IceContextCreateWithCredentials(const_cast<char *>(name_.c_str()),
+                                                  flags,
+                                                  const_cast<char*>(ufrag_.c_str()),
+                                                  const_cast<char*>(upass_.c_str()), &ctx_);
+  if (r) {
+    ELOG_WARN("%s message: Couldn't create ICE ctx", toLog());
+    start_promise_.set_value();
+    return;
+  }
+
+  r = nicer_->IceContextSetTrickleCallback(ctx_, &NicerConnection::trickle_callback, this);
```

Comment:
no, we don't. This is the only callback we have to get info from the local candidates. It's similar to `cb_new_candidate` in LibNice.